fixing issue with drawing elements with false style

marq24 opened the pull request:
When the style of the drawable has changed the level needs to adjusted before draw(...) is called - in any other case the you have the risk, that the false style will be applied to the drawable [since the false bucket (with wrong style) will be returned when calling t.buckets.get{.*}Bucket(level)]

In my project (GPSLogger) I had the render a path where different segments should have a different color - I came across a bug, that at certain conditions the path segments switched their color (the one that should had been blue was green and vise versa - or all segments have been just green or just blue) - depending on the order I add my LineDrawables in my modified setLineString(...) method.

At the end of the day I found that in the common base class (VectorLayer) the detection IF the style has been changed is called "too late" - it have to be called before the draw (and not after) - yes this implies the additional null check for lastStyle... As alternative you can init _level_ with -2 (which is not very intuitive)

devemux86 commented:
Thanks for the detailed info.

Do you have some short, self contained example,
so we can reproduce / verify the need of the change?

marq24 commented:
how to reproduce? Edit the org.oscim.android.test.VectorLayerActivity and insert this before 

`vectorLayer.update();`

```
final GeometryFactory geomFactory = new GeometryFactory(new PackedCoordinateSequenceFactory());
Style.Builder sb2 = Style.builder()
        .strokeWidth(5 * CanvasAdapter.getScale())
        .strokeColor(Color.BLUE)
        .randomOffset(false)
        .fixed(true);

double[] l1 = new double[]{8,50,9,51};
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(l1, 2), geomFactory), sb2.build()));

sb2.strokeColor(Color.GREEN);
double[] l2 = new double[]{9,51,10,52};
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(l2, 2), geomFactory), sb2.build()));
```

Without the change you will just see a complete green line above Germany: 
![Screenshot_20200802-115354](https://user-images.githubusercontent.com/8012359/89120663-62825100-d4b8-11ea-8a80-bff2651028a7.png)

Only when you pan in a way, that the second line will not be drawn (cause it will be filtered out by the `mDrawables.search(bbox, tmpDrawables);`), then the line will change it's color: 
![Screenshot_20200802-115344](https://user-images.githubusercontent.com/8012359/89120668-66ae6e80-d4b8-11ea-803d-044fcf8a3c52.png)

Once the fix is applied the two lines will be rendered based on their given style:
![Screenshot_20200802-115425](https://user-images.githubusercontent.com/8012359/89120709-c0af3400-d4b8-11ea-8147-0e18c4108e12.png)

devemux86 commented:
> Without the change you will just see a complete green line above Germany

That's strange, testing your sample the `master` code shows correctly the 2 line colors.

marq24 commented:
I am sorry - I made the sample to easy... 

so please modify the `org.oscim.android.test.VectorLayerActivity` so it will look like this... When you then zoom to that "path" section in France (near Langres) you hopefully will see the bug - when you zoom and pan, so that the path will be clipped, then parts of the path are going to change their colors... 

Important here you need to remove the "PointDrawable" part when using your original master branch - since in any other case you get an IllegalArgumentException (at org.oscim.renderer.bucket.RenderBuckets.getBucket(RenderBuckets.java:210)) [this issue will be also fixed with this change]

```
/*Style.Builder sb = Style.builder()
        .buffer(0.5)
        .fillColor(Color.RED)
        .fillAlpha(0.2f);

for (int i = 0; i < 2000; i++) {
    Style style = sb.buffer(Math.random() + 0.2)
            .fillColor(ColorUtil.setHue(Color.RED,
                    (int) (Math.random() * 50) / 50.0))
            .fillAlpha(0.5f)
            .build();

    vectorLayer.add(new PointDrawable(Math.random() * 180 - 90,
            Math.random() * 360 - 180,
            style));

}*/

final GeometryFactory geomFactory = new GeometryFactory(new PackedCoordinateSequenceFactory());
Style s1 = Style.builder()
        .strokeWidth(3 * CanvasAdapter.getScale())
        .strokeColor(Color.BLUE)
        .randomOffset(false)
        .fixed(true).build();

Style s2 = Style.builder()
        .strokeWidth(3 * CanvasAdapter.getScale())
        .strokeColor(Color.GREEN)
        .randomOffset(false)
        .fixed(true).build();

vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.528778333333334,48.00876166666667,5.5287733333333335,48.008765}, 2), geomFactory), s1));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.52898,48.00906666666666,5.528945,48.00987833333333}, 2), geomFactory), s1));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.528698333333334,48.00991333333333,5.522058333333333,48.02002833333333}, 2), geomFactory), s1));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.52201,48.02027833333333,5.52201,48.02031,5.518418333333333,48.027506666666675}, 2), geomFactory), s1));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.518321666666666,48.027550000000005,5.5135700000000005,48.031378333333336}, 2), geomFactory), s1));

vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.5287733333333335,48.008765,5.52898,48.00906666666666}, 2), geomFactory), s2));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.528945,48.00987833333333,5.528698333333334,48.00991333333333}, 2), geomFactory), s2));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.522058333333333,48.02002833333333,5.52201,48.02027833333333}, 2), geomFactory), s2));
vectorLayer.add(new LineDrawable(new LineString(new PackedCoordinateSequence.Double(new double[]{5.518418333333333,48.027506666666675,5.518321666666666,48.027550000000005}, 2), geomFactory), s2));

```

devemux86 commented:
Thanks, merged via 8602b5780a486644441600958333365a55edc76d.
